File: 0049-group-anagrams/0049-group-anagrams.py
class Solution(object):
#Creates a dictionary except if the sorted version of the word does not exist yet it will auto create it with a default word
#This avoids having to cheack if that key already exist in the dictionary 
    from collections import defaultdict
    def groupAnagrams(self, strs):

        #Create a bucket organizer so every key will map to a list of words 

        AnagramMap = defaultdict(list)

        #Vist every word in the input lisst one at a time 

        for word in strs:

            #Take the word and return the letters in alphabetical order as a list

            #Join the list of all the letters back 

            key = "".join(sorted(word))

            #Place the orignial word into the bucket labelled by its sorted key

            AnagramMap[key].append(word)

        return list(AnagramMap.values())


        

File: 0049-group-anagrams/0049-group-anagrams.java
import java.util.*;

class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        // HashMap stores:
        // key   = sorted version of the word
        // value = list of words that are anagrams
        Map<String, List<String>> result = new HashMap<>();

        // Go through every word in the input
        for (String word : strs) {

            // Convert the word into a character array
            char[] chars = word.toCharArray();

            // Sort the characters
            // Example: "eat" -> ['a', 'e', 't']
            // Anagrams will have the same sorted characters
            Arrays.sort(chars);

            // Convert the sorted characters back into a String
            // This becomes our key
            String key = new String(chars);

            // If this key doesn't exist, create a new ArrayList
            // Then add the original word to that list
            result.computeIfAbsent(key, k -> new ArrayList<>()).add(word);
        }

        // Return only the lists of anagrams
        return new ArrayList<>(result.values());
    }
}